framework/src/media/audio/audio_manager.c

```diff
@@ -528,24 +606,53 @@ int start_audio_stream_out(void *data, uint32_t frames)
 	int retry = AUDIO_STREAM_RETRY_COUNT;
 	struct pcm *pcm;
 
+	// Todo: Need to check g_actual_audio_out_card_id index.
+	audio_card_info_t *cur_card = &g_audio_out_cards[g_actual_audio_out_card_id];
+	auddbg("start_audio_stream_out\n");
+
 	if ((ret = get_active_audio_device_pcm(&pcm, OUTPUT)) != AUDIO_MANAGER_SUCCESS) {
+		auddbg("God damn error\n");
 		return ret;
 	}
 
-	if (g_audio_out_cards[g_actual_audio_out_card_id].status == AUDIO_CARD_PAUSE) {
+	if (cur_card->status == AUDIO_CARD_PAUSE) {
 		ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_RESUME, NULL);
 		if (ret < 0) {
 			auddbg("Fail to ioctl AUDIOIOC_RESUME, ret = %d\n", ret);
 			return AUDIO_MANAGER_DEVICE_FAIL;
 		}
 
-		g_audio_out_cards[g_actual_audio_out_card_id].status = AUDIO_CARD_RUNNING;
-		audvdbg("Resume the output audio card!!\n");
+		cur_card->status = AUDIO_CARD_RUNNING;
+		auddbg("Resume the output audio card!!\n");
 	}
 
 	do {
-		audvdbg("Start Playing!!\n");
-		ret = pcm_writei(pcm, data, frames);
+		auddbg("Start Playing!!\n");
+
+		if(cur_card->resample.necessary) {
+			src_data_t srcData;
+            srcData.data_in = data;
+            srcData.input_frames = frames;
+            srcData.channels_num = 2; // Todo: Consider Mono/Stereo
+			srcData.origin_sample_rate = cur_card->resample.from;
+			srcData.origin_sample_width = SAMPLE_WIDTH_16BITS;
+            srcData.desired_sample_rate = cur_card->resample.to;
+			srcData.desired_sample_width = SAMPLE_WIDTH_16BITS;
+			srcData.input_frames_used = 0;
+
+			//while(done < (int)((float)frames * cur_card->resample.ratio)) {
+				if(src_simple(cur_card->resample.handle, &srcData) != SRC_ERR_NO_ERROR) {
+					auddbg("Fail to resample to %u from %u\n", srcData.desired_sample_rate, srcData.origin_sample_rate);
+					return AUDIO_MANAGER_RESAMPLE_FAIL;
+				}
+
+				pcm_writei(pcm, srcData.data_out, srcData.output_frames_gen);
```

Comment:
same as above, 3rd param of writei is frame, if output_frames_gen contains number of channel, below is correct.
pcm_writei(pcm, srcData.data_out, (srcData.output_frames_gen / channel));